core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -286,6 +286,7 @@ class DAGScheduler(
       case None =>
         // We are going to register ancestor shuffle dependencies
         getAncestorShuffleDependencies(shuffleDep.rdd).foreach { dep =>
+          assert(!shuffleToMapStage.get(dep.shuffleId).isDefined)
```

Comment:
Why are you asserting this when you are already within a match case where this must be true?  If it is actually possible for the contents of shuffleToMapStage to change between the pattern match and the foreach, then we need to fix the unsynchronized update, not just shutdown Spark on a failed assertion.